internal/component/database_observability/mysql/collector/explain_plan.go

```
@@ -0,0 +1,750 @@
+	}
+	output.Plan = planNode
+
+	return output, nil
+}
+
+func parseTopLevelPlanNode(logger log.Logger, topLevelPlanNode []byte) (planNode, error) {
+	// Table at top level
+	if table, _, _, err := jsonparser.Get(topLevelPlanNode, "table"); err == nil {
+		tableDetails, err := parseTableNode(logger, table)
+		if err != nil {
+			return planNode{}, err
+		}
+
+		return tableDetails, nil
+	} // else we don't find the table at the top level. This is likely not an error because the NodeDetails may be for a different type of operation.
+	// We may have to check to see if no operation has been set after checking all known operation types.
+
+	// Nested Loop Join
+	if nestedLoopJoin, _, _, err := jsonparser.Get(topLevelPlanNode, "nested_loop"); err == nil {
+		pnode, err := parseNestedLoopJoinNode(logger, nestedLoopJoin)
+		if err != nil {
+			return planNode{}, err
+		}
+		return pnode, nil
+	}
+
+	// Grouping Operation
+	if groupingOperation, _, _, err := jsonparser.Get(topLevelPlanNode, "grouping_operation"); err == nil {
+		pnode, err := parseGroupingOperationNode(logger, groupingOperation)
+		if err != nil {
+			return planNode{}, err
+		}
+		return pnode, nil
+	}
+
+	// Ordering Operation
+	if orderingOperation, _, _, err := jsonparser.Get(topLevelPlanNode, "ordering_operation"); err == nil {
+		pnode, err := parseOrderingOperationNode(logger, orderingOperation)
+		if err != nil {
+			return planNode{}, err
+		}
+		return pnode, nil
+	}
+
+	// Duplicates Removal
+	if duplicatesRemoval, _, _, err := jsonparser.Get(topLevelPlanNode, "duplicates_removal"); err == nil {
+		pnode, err := parseDuplicatesRemovalNode(logger, duplicatesRemoval)
+		if err != nil {
+			return planNode{}, err
+		}
+		return pnode, nil
+	}
+
+	if unionResult, _, _, err := jsonparser.Get(topLevelPlanNode, "union_result"); err == nil {
+		pnode, err := parseUnionResultNode(logger, unionResult)
+		if err != nil {
+			return planNode{}, err
+		}
+		return pnode, nil
+	}
+
+	return planNode{}, nil
```

What happens here? We didn't understand what the top level node was, should we throw an error?

---

So, a couple of things are at play here.

First, "top level node" is perhaps misleading. We're actually looking for what type of operation this is, usually encapsulated in a top level query_block. However this get's called for other node/operation types which have nested operations, and those operations could, theoretically be any of the "top level" nodes.

See `parseGroupingOperationNode` through `parseUnionResultNode` where [this method is called](https://github.com/grafana/alloy/blob/b3bb8a3ae50376db9a115de313d5f0f1046ee1cb/internal/component/database_observability/mysql/collector/explain_plan.go#L345).

Second, we're only parsing parts of the explain plan which we *do* know about. It's entirely likely that there are properties which we *don't* know about. The result would be an empty node in the ExplainPlanOutput struct, and is a good reason why we still need the debug output of the raw explain plan, and the proposed `TODO` of providing context of [where errors occur](https://github.com/grafana/alloy/blob/b3bb8a3ae50376db9a115de313d5f0f1046ee1cb/internal/component/database_observability/mysql/collector/explain_plan.go#L679).

---

So perhaps the final `planNode` here should represent something unknown then in this case?
Like
`planNode{Operation: "unknown"}`

---

Ahh yes, I like this. I've added it [here](https://github.com/grafana/alloy/pull/3801/commits/74be1ac7f4bab9417aeff72cf52eae4c9a47a414).